sdk/src/utils.ts

```diff
@@ -750,3 +770,158 @@ export const getMinTick = (tickSpacing: number) => {
+  const sqrtPrice = Math.sqrt(previousSqrtPrice * currentSqrtPrice)
+  return volumeX + volumeY / sqrtPrice
+}
+
+// export const getRangeFromPrices = (
+//   priceMinTokenX: number,
+//   priceMaxTokenX: number,
+//   priceMinTokenY: number,
+//   priceMaxTokenY: number,
+//   tickmap: Tickmap,
+//   tickSpacing: number
+// ): LiquidityRange => {
+//   // calculate ticks from prices ratio between A and B
+//   let tickLower = priceToTick(priceMinTokenY / priceMinTokenX)
+//   let tickUpper = priceToTick(priceMaxTokenY / priceMaxTokenX)
+
+//   // check if tick are initialized
+//   const isLowerInitialized = isInitialized(tickmap, tickLower, tickSpacing)
+//   const isUpperInitialized = isInitialized(tickmap, tickUpper, tickSpacing)
+
+//   //find accurate if not initialized
+//   if (!isLowerInitialized) {
+//     //find closest tick on the right
+//     const closestTicks = findClosestTicks(
+//       tickmap.bitmap,
+//       tickLower,
+//       tickSpacing,
+//       MAX_TICK,
+//       tickUpper,
+//       'up'
+//     )
+//     if (!closestTicks.length) {
+//       throw new Error(Errors.TickNotFound)
+//     }
+//     tickLower = closestTicks[0]
+//   }
+
+//   if (!isUpperInitialized) {
+//     //find closest tick on the left
+//     const closestTicks = findClosestTicks(
+//       tickmap.bitmap,
+//       tickUpper,
+//       tickSpacing,
+//       MIN_TICK,
+//       tickLower,
+//       'down'
+//     )
+//     if (!closestTicks.length) {
+//       throw new Error(Errors.TickNotFound)
+//     }
+//     tickUpper = closestTicks[0]
+//   }
+
+//   return { tickLower, tickUpper }
+// }
+
+export const getTickArray = (rawTicks: Tick[]): ParsedTick[] => {
+  const sortedTicks = rawTicks.sort((a, b) => a.index - b.index)
+  const ticks: ParsedTick[] = rawTicks.length ? parseFeeGrowthAndLiquidityOnTicks(sortedTicks) : []
+  if (!ticks.length) {
+    throw new Error(Errors.TickArrayIsEmpty)
+  }
+  return ticks
+}
+
+export const getTokenXInRange = (ticks: ParsedTick[], lowerTick: number, upperTick: number): BN => {
+  let sumTokenX: BN = new BN(0)
+  let currentIndex = 0
+  let nextIndex = 0
+
+  for (let i = 0; i < ticks.length - 1; i++) {
+    currentIndex = ticks[i].index
+    nextIndex = ticks[i + 1].index
+
+    if (currentIndex >= lowerTick && currentIndex < upperTick) {
+      const lowerSqrtPrice = calculatePriceSqrt(currentIndex)
+      const upperSqrtPrice = calculatePriceSqrt(nextIndex)
+
+      sumTokenX = sumTokenX.add(
+        getXfromLiquidity(ticks[i].liquidity, upperSqrtPrice.v, lowerSqrtPrice.v)
+      )
+    }
+  }
+  return sumTokenX
+}
+
+export const getRangeBasedOnFeeGrowth = (
+  tickArrayPrevious: ParsedTick[],
+  tickArrayCurrent: ParsedTick[]
+) => {
+  let tickLower: number = null
+  let tickUpper: number = null
+  let tickLowerSaved = false
+  let lastIndex = 0
+
+  for (let i = 0; i < tickArrayPrevious.length - 1; i++) {
+    let previousTick = tickArrayPrevious[i]
+    let currentTick = tickArrayCurrent[i]
```

Comment:
this for sure should't work. Find current tick state based on the same index. If current tick state not exist - don not mark this tick as changed